hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/CloseRegionHandler.java

```diff
@@ -52,85 +54,101 @@
-      final RegionInfo regionInfo, final boolean abort,
-      ServerName destination) {
-    this(server, rsServices, regionInfo, abort,
-      EventType.M_RS_CLOSE_REGION, destination);
+  public CloseRegionHandler(final RegionServerServices rsServices, final RegionInfo regionInfo,
+    final boolean abort, @Nullable ServerName destination) {
+    this(rsServices, regionInfo, abort, EventType.M_RS_CLOSE_REGION, destination);
   }
 
-  protected CloseRegionHandler(final Server server,
-      final RegionServerServices rsServices, RegionInfo regionInfo,
-      boolean abort, EventType eventType, ServerName destination) {
-    super(server, eventType);
-    this.server = server;
-    this.rsServices = rsServices;
+  protected CloseRegionHandler(RegionServerServices rsServices, RegionInfo regionInfo,
+      boolean abort, EventType eventType, @Nullable ServerName destination) {
+    super(rsServices, eventType);
     this.regionInfo = regionInfo;
     this.abort = abort;
+    this.rss = rsServices;
     this.destination = destination;
   }
 
-  public RegionInfo getRegionInfo() {
-    return regionInfo;
-  }
-
-  @Override
-  public void process() {
+  @Override public void process() throws IOException {
     try {
-      String name = regionInfo.getEncodedName();
-      LOG.trace("Processing close of {}", name);
-      String encodedRegionName = regionInfo.getEncodedName();
       // Check that this region is being served here
-      HRegion region = (HRegion)rsServices.getRegion(encodedRegionName);
+      HRegion region = (HRegion)this.rss.getRegion(this.regionInfo.getEncodedName());
       if (region == null) {
-        LOG.warn("Received CLOSE for region {} but currently not serving - ignoring", name);
+        LOG.warn("Received CLOSE for {} but not ONLINE; ignoring CLOSE request",
+          this.regionInfo.getRegionNameAsString());
         // TODO: do better than a simple warning
         return;
       }
-
-      // Close the region
-      try {
-        if (region.close(abort) == null) {
-          // This region got closed.  Most likely due to a split.
-          // The split message will clean up the master state.
-          LOG.warn("Can't close region {}, was already closed during close()", name);
-          return;
-        }
-      } catch (IOException ioe) {
-        // An IOException here indicates that we couldn't successfully flush the
-        // memstore before closing. So, we need to abort the server and allow
-        // the master to split our logs in order to recover the data.
-        server.abort("Unrecoverable exception while closing region " +
-          regionInfo.getRegionNameAsString() + ", still finishing close", ioe);
-        throw new RuntimeException(ioe);
+      if (region.getCoprocessorHost() != null) {
+        // XXX: The behavior is a bit broken. At master side there is no FAILED_CLOSE state, so if
+        // there are exceptions thrown from the CP, we can not report the error to the master, and
```

Comment:
😱

Reply:
Sorry, i'm at reviewer fatigue at this point. Will be back.

Reply:
Thanks Nick